### ISSUE TITLE
PV (voltage controlled generator) as slack bus and distributed slack bus

### ISSUE THREAD

**Issue by TonyXiang8787** [PowerGridModel/power-grid-model#185]
https://github.com/e2nIEE/pandapower/issues/771
https://github.com/e2nIEE/pandapower/issues/1399
https://github.com/e2nIEE/pandapower/blob/develop/tutorials/distributed_slack.ipynb

https://github.com/powsybl/powsybl-open-loadflow

**Comment by TonyXiang8787** [comment]
@sudo-ac yes, this issue is meant to define PV nodes. However, PV nodes rarely (to none) exist in distribution grids. This project is becoming more fine-tuned for DSOs. So this issue is likely to be closed.

**Comment by TonyXiang8787** [comment]
> Do you mean that you will never implement PV nodes in PGM?

Unless there is a major shift of the focus of PGM, it is unlikely to implement PV nodes.

**Comment by petersalemink95** [comment]
@sudo-ac thank you for the interest in this issue. Since the focus of the Power Grid Model project is mainly on distribution systems it is unlikely that PV nodes will be included in Power Grid Model, unless there comes a shift in focus. Therefore I am closing this issue.

**Comment by petersalemink95** [comment]
@scud-soptim I am re-opening the issue. The focus of Power Grid Model will still be on distribution grids, but we are open for transmission components if they fit within the current framework of PGM. So, to you, or anyone interested: feel free to implement this in PGM.

**Comment by scud-soptim** [comment]
@petersalemink95 Thanks, maybe we find the time for it, it would be awesome to have that feature.

**Comment by petersalemink95** [comment]
@scud-soptim That would be great! Let us know if you do - we'd be happy to think along!

**Comment by petersalemink95** [comment]
Writing previously agreed implementation details, for reviewing purposes:
- PV bus is supported for `Newton-Raphson` power flow
- PV buses are implemented by adding a new component class: `VoltageRegulator`. The `voltage_regulator` can be added by a user next to a `load` / `gen` component. When both a `voltage_regulator` regulates a `load` or `gen`  the node will function as a PV node: `p_specified` coming from the `load` / `gen` and `u_ref` coming from the `voltage_regulator`
- `VoltageRegulator` attributes: `id`, `controlled_object`, `u_ref`, `q_min`, `q_max`
- Support both `sym` and `asym` calculations. For both cases there should be a single `u_ref`: target voltage should be the same across all phases
- We keep the source as is, functioning as a slack bus
- Q limits are included
- We support multiple voltage controlled generators on 1 node if they have the same target voltage
- We include some sort of cooldown factor, to avoid going back and forth between PV and PQ buses (when operating at Q limits the bus will behave as a PQ bus instead of PV bus)
- For power flow with automatic tap regulation we check if it works as expected with PV buses. If not, we do not allow PV buses in this calculation
- We keep the matrix structure the same

**Comment by mgovers** [comment]
This issue was accidentally closed by a linked PR (#1220). While that PR does contain an implementation for this issue, it is still behind an experimental feature flag. Re-opening for now until the experimental feature flag is removed
